File: kattis/open/primecount/primecount.py
def solve():
    n = int(input())

    if n < 2:
        print(0)
        exit()

    s = int(n ** 0.5)
    vals = []
    i = 1

    while i <= n:
        v = n // i
        i = n // v + 1
        vals.append(v)

    m = len(vals)
    id1 = [0] * (s + 1)
    id2 = [0] * (s + 1)

    for i, v in enumerate(vals):
        if v <= s:
            id1[v] = i
        else:
            id2[n // v] = i

    g = [v - 1 for v in vals]

    for p in range(2, s + 1):
        if g[id1[p]] == g[id1[p - 1]]:
            continue

        sp = g[id1[p - 1]]
        p2 = p * p

        for i in range(m):
            v = vals[i]
            if v < p2:break

            div = v // p
            idx = id1[div] if div <= s else id2[n // div]
            g[i] -= g[idx] - sp

    print(g[0])
solve()

File: kattis/open/primecount/primecount.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>

using namespace std;

void solve() {
    long long n;
    cin >> n;

    if (n < 2) {
        cout << 0 << endl;
        exit(0);
    }

    long long s = sqrt(n);
    vector<long long> vals;
    long long i = 1;

    while (i <= n) {
        long long v = n / i;
        i = n / v + 1;
        vals.push_back(v);
    }

    int m = vals.size();
    vector<int> id1(s + 1);
    vector<int> id2(s + 1);

    for (int i = 0; i < m; ++i) {
        long long v = vals[i];
        if (v <= s) {
            id1[v] = i;
        } else {
            id2[n / v] = i;
        }
    }

    vector<long long> g;
    g.reserve(m);
    for (long long v : vals) {
        g.push_back(v - 1);
    }

    for (int p = 2; p <= s; ++p) {
        if (g[id1[p]] == g[id1[p - 1]]) {
            continue;
        }

        long long sp = g[id1[p - 1]];
        long long p2 = (long long)p * p;

        for (int i = 0; i < m; ++i) {
            long long v = vals[i];
            if (v < p2) break;

            long long div = v / p;
            int idx = (div <= s) ? id1[div] : id2[n / div];
            g[i] -= g[idx] - sp;
        }
    }

    cout << g[0] << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    solve();
    return 0;
}